File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n=s.size();
        if(n==0)
            return "";
        if(n==1)
            return s;
        
        vector<vector<bool>> dp(n+1,vector<bool>(n+1,false));
        
        for(int i=0;i<n;i++)
            dp[i][i]=true;
        
        int start=0;
        int maxlen=1;
        
        for(int i=0;i<n-1;i++)
        {
            if(s[i]==s[i+1])
            {
                dp[i][i+1]=true;
                start=i;
                maxlen=2;
            }
        }
        
        for(int len=3;len<=n;len++)
        {
            for(int i=0;i<=n-len;i++)
            {
                int j=i+len-1;
                if(s[i]==s[j] && dp[i+1][j-1])
                {
                    dp[i][j]=true;
                    if(len>maxlen)
                {start=i;
                maxlen=len;}
                }
            }
        }
        
        return s.substr(start,maxlen);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    public boolean isPalindrome(int i,int j,String s,Boolean[][] dp)
    {
        if(i>=j)
        return true;

        if(dp[i][j]!=null)
        return dp[i][j];

        return dp[i][j] = (s.charAt(i) == s.charAt(j)) && isPalindrome(i + 1, j - 1, s, dp);
        
    }

    public String longestPalindrome(String s) {
        int n=s.length();
        if(n<=1)
        return s;
        Boolean[][] dp=new Boolean[n][n];
        int start=0;
        int maxi=1;
        for(int i=0;i<n;i++)
        {
            for(int j=i;j<n;j++)
            {
                if(isPalindrome(i,j,s,dp) && (j-i+1>maxi))
                {
                    start=i;
                    maxi=j-i+1;
                }
            }
        }

        return s.substring(start,start+maxi);
    }
}